pmd-java/src/main/java/net/sourceforge/pmd/cpd/JavaTokenizer.java

```diff
@@ -170,27 +170,36 @@ protected boolean isLanguageSpecificDiscarding() {
         }
 
         private void detectAnnotations(JavaccToken currentToken) {
+            updateAnnotationStatus(currentToken);
+            handleAnnotationStack(currentToken);
+        }
+        
+        private void updateAnnotationStatus(JavaccToken currentToken) {
             if (isAnnotation && nextTokenEndsAnnotation) {
                 isAnnotation = false;
                 nextTokenEndsAnnotation = false;
             }
-            if (isAnnotation) {
-                if (currentToken.kind == JavaTokenKinds.LPAREN) {
-                    annotationStack++;
-                } else if (currentToken.kind == JavaTokenKinds.RPAREN) {
-                    annotationStack--;
-                    if (annotationStack == 0) {
-                        nextTokenEndsAnnotation = true;
-                    }
-                } else if (annotationStack == 0 && currentToken.kind != JavaTokenKinds.IDENTIFIER
-                        && currentToken.kind != JavaTokenKinds.LPAREN) {
-                    isAnnotation = false;
-                }
-            }
             if (currentToken.kind == JavaTokenKinds.AT) {
                 isAnnotation = true;
             }
         }
+        
+        private void handleAnnotationStack(JavaccToken currentToken) {
+            if (!isAnnotation) {
+                return;
```

Comment:
move return at the end of the method.
Reason: Adding a if block for `!isAnnotation` is a good practice but increases cylomatic complexity of the method. Also, there is no specific operation done for `!isAnnotation` so the block becomes redundant. Rather add return at the end of the method is needed for developers understandability. or you can call handleAnnotationStack only if `isAnnotation == Boolean.TRUE` in detectAnnotation method